### PR TITLE
feat: add LinkIndexer and CI

### PR DESCRIPTION
**Pull request by olizilla**
- Port LinkIndexer from https://github.com/web3-storage/web3.storage/pull/1785
- Add tests, lint and type checking
- Add type generated type declarations
- Add CI

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -20,3 +32,110 @@ export async function * linkdex (carStream) {
...
+      this._index(block)
+    }
+  }
+
+  /**
+   * Index all the links from the block
+   * @param {import('multiformats/block').Block} block
+   */
+  _index (block) {
+    this.indexed++
```
Comment (vasco-santos) [inline index.js:74]: Should this be after next condition where we validation if given block was already indexed?

**Review by vasco-santos — CHANGES_REQUESTED** [review]
Great work, I like the new `LinkIndexer` :)

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,9 @@
+dist/*
+!dist/index*
+!dist/decode*
```
[the remark this answers is not in the document]
Comment (olizilla) [reply .gitignore:3]: ah, so I'm consuming this module direct from GitHub at the moment... but yes, I should probably just publish it and remove the dist.

File: index.js (hunk 1 above)
Comment (olizilla) [reply index.js:74]: I'm using indexed to count how may blocks this thing was asked to index... so a total block count. Later we use the map size to find the unique block count.

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (vasco-santos) [reply .gitignore:3]: If you want to iterate like this for now, I am ok on getting the release job once you are done

**Review by vasco-santos — COMMENTED** [review]
You can iterate further without release job if it is helpful. We can remove the dist bits and add the release job once you are done

File: package.json [hunk 3]
```diff
@@ -4,12 +4,16 @@
...
+  "types": "./dist/index.d.ts",
   "bin": {
     "linkdex": "./bin.js"
   },
   "scripts": {
     "start": "./bin.js",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "ava",
+    "lint": "standard",
+    "tsc": "tsc",
```
[the remark this answers is not in the document]
Comment (olizilla) [reply package.json:15]: @vasco-santos is that so we can run `npm test` and get the tsc output as well as the test output?

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (olizilla) [reply .gitignore:3]: I think I can use a `prepare` script to have it run tsc before publish or when installed via git: https://docs.npmjs.com/cli/v8/using-npm/scripts#life-cycle-scripts

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (vasco-santos) [reply package.json:15]: yeah, exactly. I always add it to test as `tsc` compile should also succeed for us to push the commit and while developing we likely won't remember to run it

[the remark this answers is not in the document]
Comment (olizilla) [reply package.json:19]: Done!

**Pull request merged**